File: packages/oxlint-plugin-react-doctor/src/plugin/rules/tanstack-start/tanstack-start-no-dynamic-server-fn-import.ts
import { TANSTACK_SERVER_FN_FILE_PATTERN } from "../../constants/tanstack.js";
import { defineRule } from "../../utils/define-rule.js";
import type { Rule } from "../../utils/rule.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";

export const tanstackStartNoDynamicServerFnImport = defineRule<Rule>({
  id: "tanstack-start-no-dynamic-server-fn-import",
  title: "Dynamic server function import",
  tags: ["test-noise"],
  requires: ["tanstack-start"],
  severity: "error",
  recommendation:
    "Use `import { myFn } from '~/utils/my.functions'`. The bundler only swaps server code for RPC stubs on static imports.",
  create: (context: RuleContext) => ({
    ImportExpression(node: EsTreeNodeOfType<"ImportExpression">) {
      const source = node.source;
      if (!source) return;

      let importPath: string | null = null;
      if (isNodeOfType(source, "Literal") && typeof source.value === "string") {
        importPath = source.value;
      } else if (isNodeOfType(source, "TemplateLiteral") && source.quasis?.length === 1) {
        importPath = source.quasis[0].value?.raw ?? null;
      }

      if (importPath && TANSTACK_SERVER_FN_FILE_PATTERN.test(importPath)) {
        context.report({
          node,
          message:
            "Dynamically importing a server-functions file leaks server code into the client bundle.",
        });
      }
    },
  }),
});

File: native/oxlint/rules/tanstack-start-no-dynamic-server-fn-import.rs
use oxc_ast::{ast::Expression, AstKind};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;

use crate::{
    context::{ContextHost, LintContext},
    rule::Rule,
    AstNode,
};

const MESSAGE: &str =
    "Dynamically importing a server-functions file leaks server code into the client bundle.";
const SERVER_FUNCTION_FILE_SUFFIXES: [&str; 5] = [
    ".functions",
    ".functions.js",
    ".functions.jsx",
    ".functions.ts",
    ".functions.tsx",
];

#[derive(Debug, Default, Clone)]
pub struct TanstackStartNoDynamicServerFnImport;

declare_oxc_lint!(
    /// Disallow dynamic imports of TanStack Start server-function modules.
    TanstackStartNoDynamicServerFnImport,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Disallow dynamic server-function imports.",
);

impl Rule for TanstackStartNoDynamicServerFnImport {
    fn should_run(&self, ctx: &ContextHost) -> bool {
        !is_non_production_file(ctx)
    }

    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let AstKind::ImportExpression(import_expression) = node.kind() else {
            return;
        };
        let import_path = match &import_expression.source {
            Expression::StringLiteral(string_literal) => string_literal.value.as_str(),
            Expression::TemplateLiteral(template_literal) if template_literal.quasis.len() == 1 => {
                template_literal.quasis[0].value.raw.as_str()
            }
            _ => return,
        };
        if !SERVER_FUNCTION_FILE_SUFFIXES
            .iter()
            .any(|suffix| import_path.ends_with(suffix))
        {
            return;
        }
        ctx.diagnostic(OxcDiagnostic::warn(MESSAGE).with_label(import_expression.span));
    }
}
